refactor(tagfilter): improve tag expression parser with better logic …

Groovin-Dev opened the pull request:
…and debugging

This commit completely refactors the tag filtering system used in storage and export buses. Key improvements include:

- Replace the old match rule system with a proper expression tree
- Add comprehensive DEBUG_ENABLED logging throughout the codebase
- Fix issues with AND expressions by implementing specialized handling
- Improve handling of edge cases and invalid expressions
- Add better support for parentheses and operator precedence
- Implement more robust wildcard pattern matching using regex

These changes significantly improve the robustness of the tag filtering system while maintaining backward compatibility with existing configurations.

Groovin-Dev committed:
refactor(tagfilter): improve tag expression parser with better logic and debugging

This commit completely refactors the tag filtering system used in storage
and export buses. Key improvements include:

- Replace the old match rule system with a proper expression tree
- Add comprehensive DEBUG_ENABLED logging throughout the codebase
- Fix issues with AND expressions by implementing specialized handling
- Improve handling of edge cases and invalid expressions
- Add better support for parentheses and operator precedence
- Implement more robust wildcard pattern matching using regex

These changes significantly improve the robustness of the tag filtering
system while maintaining backward compatibility with existing configurations.

GlodBlock commented on src/main/java/com/glodblock/github/extendedae/common/me/taglist/TagPriorityList.java:
```diff
@@ -45,27 +68,174 @@ public Iterable<AEKey> getItems() {
+        }
+        
+        // Special case: Detect if we're dealing with an AND expression with direct tag expressions
+        if (parsedExpression instanceof TagExpParser.AndExpression andExpression) {
```
What will happen if it is a nested AND expression? like `(exp1 & exp2) | (exp3 & exp4)`

Groovin-Dev replied:
@GlodBlock it works now :) Just pushed change to update that and have more robust support.

Input chest:
![image](https://github.com/user-attachments/assets/918eafdc-b867-4f2c-bbf4-f76594d19491)
![image](https://github.com/user-attachments/assets/3ac3bcbf-d3cf-4dee-9230-b742db6d6d3c)

Filter (`(c:tools) & !(c:tools/paxel)`):
![image](https://github.com/user-attachments/assets/2b019311-c063-4a8e-b6ba-1644554eaf0c)


Output:
![image](https://github.com/user-attachments/assets/5dd0f12b-7737-447e-bb76-3b3e23a22c1b)

Groovin-Dev committed:
fix(taglist): improve tag expression parser for wildcards and namespaces

- Fix tokenization validation to properly handle wildcards (*) and namespaces (:)
- Replace dynamic programming wildcard matching with regex-based pattern matching
- Properly validate tag tokens at end of expressions to prevent parsing errors
- Eliminate "Expected operand after last token" errors when using patterns like
  'allthecompressed*' in Tag Export Bus and Tag Storage Bus

Fixes parsing errors reported in application logs when using mod tags with wildcards.

Groovin-Dev committed:
Merge branch '1.21.1-neoforge'

# Conflicts:
#	src/main/java/com/glodblock/github/extendedae/common/me/taglist/TagExpParser.java
#	src/main/java/com/glodblock/github/extendedae/common/me/taglist/TagPriorityList.java
#	src/main/java/com/glodblock/github/extendedae/common/parts/PartTagExportBus.java
#	src/main/java/com/glodblock/github/extendedae/common/parts/PartTagStorageBus.java

GlodBlock commented:
LGTM
